docs/SPEC.md

```diff
@@ -839,6 +833,32 @@ Side effects can occur in two ways.
 
 A function produces side effects when it is explicitly declared to have side effects or when it calls a function that itself produces side effects.
 
+## Package initialization
+
+A package with no imports is initialized by first initializing and assigning all declared options and then subsequently initializing and assigning all declared variables.
+In particular all package options are initialized before any variables.
+Options cannot depend on any variables declared in the same package block.
+If a package has imports, all imported packages are initialized before any options or variables.
+
+### Variable initialization
+
+Variables are initialized in declaration order but only after all of the variables they depend on have been initialized.
+For packages consisting of multiple source files, declaration order depends on the order in which the files are presented to an interpreter.
+However package variables are initialized with the same value regardless of said file order.
+A variable cannot have a direct or indirect dependency on itself.
+
+### Option initialization
+
+Options are initialized in declaration order.
+An option cannot have dependencies on any other options assigned in the same package block.
+
+### Package imports
+
+Imported packages are initialized in declaration order.
+A package cannot be imported twice in the same file block.
+A package will only be initialized once even if it is imported multiple times in a package.
+A package may be imported purely for its side effects however any such package cannot be imported twice within the same package.
+A package cannot import two separate packages that modify the same option.
```

Comment:
This part is not quite true. We want to be able to import packages that modify the same option. 

  I think we should not be so specific about the restrictions on package imports rather state that any package imports must be defined in an unambiguous manner. 

So instead of saying that packages cannot be imported twice for side effects etc, we simply state that package imports must be deterministic. Then define what we mean by deterministic, i.e. for any packages that modify the same options their order is known. Then define the order as known for import in the same file but not across files.